src/integration/cluster_integration_tests.md

```diff
@@ -0,0 +1,210 @@
+ *	// WriteProm writes a prometheus metric. Takes tags/labels as a map for convenience.
+ *  WriteProm(name string, tags map[string]string, samples []prompb.Sample, headers Headers) error
+ *  
+ */
+_ := coordinator.WriteProm("foo_metric", map[string]string{
+	"bar_label":            "baz",
+}, []prompb.Sample{
+	{
+		Value:     42,
+		Timestamp: storage.TimeToPromTimestamp(xtime.Now()),
+	},
+}, nil)
+
+// Read some data
+/*
+ * Using this method on the resources.Coordinator interface:
+ *
+ *  // RangeQuery runs a range query with provided headers
+ *  RangeQuery(req RangeQueryRequest, headers Headers) (model.Matrix, error)
+ *  
+ */
+result, err := coord.RangeQuery(
+	resources.RangeQueryRequest{
+		Query: "foo_metric",
+		Start: time.Now().Add(-30 * time.Second),
+		End:   time.Now(),
+	},
+	nil)
+
+```
+NB: If using this code in tests, the `RangeQuery` may need to be retried to yield a result. Writes in M3 are async by default, so they're not immediately available for reads. `resources.Retry` is provided for convenience to assist with this.
+
+### Spinning up an external resource
+Occasionally, it is convenient to test M3 alongside some component it works closely with. Prometheus is the most common example. The cluster integration test framework supports this by spinning up external resources in docker containers. External resources must adhere to the following interface:
+
+```golang
+// ExternalResources represents an external (i.e. non-M3)
+// resource that we'd like to be able to spin up for an
+// integration test.
+type ExternalResources interface {
+	// Setup sets up the external resource so that it's ready
+	// for use.
+	Setup() error
+
+	// Close stops and cleans up all the resources associated with
+	// the external resource.
+	Close() error
+}
+```
+
+Since it's so commonly used in conjunction with M3, an implementation for [Prometheus](https://github.com/m3db/m3/blob/master/src/integration/resources/docker/prometheus.go) already exists.
+
+Here's an example of a test spinning up M3 and Prometheus:
+
+```golang
+cfgs, _ := inprocess.NewClusterConfigsFromConfigFile(pathToDBCfg, pathToCoordCfg, "")
+m3, _ := inprocess.NewCluster(cfgs,
+        resources.ClusterOptions{
+                DBNode: resources.NewDBNodeClusterOptions(),
+        },
+)
+defer m3.Cleanup()
+
+// Spin up external resources. In this case, prometheus.
+pool, _ := dockertest.NewPool("")
+
+prom := docker.NewPrometheus(docker.PrometheusOptions{
+	Pool:      pool,
+	PathToCfg: pathToPromCfg,
+})
+prom.Setup()
+defer prom.Close()
+
+// Run tests...
+
+
+```
+
+### Spinning up an individual component
+We've been mostly referring to a cluster with a coordinator, db node, and potentially an aggregator, but it's also possible to spin up each component individually. Each component has the same handful of constructors in the `inprocess` [package](https://github.com/m3db/m3/tree/master/src/integration/resources/inprocess). Here's an example of starting up a DB node.
```

Comment:
when would you do this out of curiosity?

Reply:
Most likely, you're debugging something in the write path of the DB node and you know you don't need a coordinator. So you could just spin up a DB node and hit it via the node APIs.